### PR TITLE
Fix bugs around docker start retries

### PR DESCRIPTION
**Pull request by notnoop**
Noticed that docker driver attempts to retry container creation and start.

This fixes a couple of bugs, where we either don't retry transient errors, or we retry in such away that makes retry invocation fails.

### CODE REVIEW

File: drivers/docker/driver.go [hunk 1]
```diff
@@ -456,6 +462,10 @@ CREATE:
 		// There is still a very small chance this is possible even with the
 		// coordinator so retry.
 		return nil, nstructs.NewRecoverableError(createErr, true)
+	} else if isDockerTransientError(createErr) && attempted < 5 {
+		attempted++
+		time.Sleep(1 * time.Second)
+		goto CREATE
 	}
 
 	return nil, recoverableErrTimeouts(createErr)
```
Comment (notnoop) [inline drivers/docker/driver.go]: It's not obvious to me why `recoverableErrTimeout` are used.  `StartTask` doesn't special case them, and wraps them before returning.  I considered removing it completely and doing more retries locally, but decided to be conservative and double check goal before making further changes to driver.

File: drivers/docker/driver.go [hunk 2]
```diff
@@ -456,6 +462,10 @@ CREATE:
 		// There is still a very small chance this is possible even with the
 		// coordinator so retry.
 		return nil, nstructs.NewRecoverableError(createErr, true)
+	} else if isDockerTransientError(createErr) && attempted < 5 {
+		attempted++
+		time.Sleep(1 * time.Second)
```
Comment (jippi) [inline drivers/docker/driver.go]: shouldn't we exponentially backoff here?

Comment (nickethier) [reply drivers/docker/driver.go]: :+1: or even just `time.Sleep(attempted * time.Second)`

File: drivers/docker/driver.go (hunk 1 above)
Comment (nickethier) [reply drivers/docker/driver.go]: I don't have much context to add here other than that it is was brought over from pre-0.9 drivers:
https://github.com/hashicorp/nomad/blob/fcc4149c55399eb4979cd3fe3fb983cfe6c8449a/client/driver/docker.go#L1701

File: drivers/docker/driver.go [hunk 3]
```diff
@@ -468,14 +478,13 @@ func (d *Driver) startContainer(c *docker.Container) error {
...
 	}
 
 	d.logger.Debug("failed to start container", "container_id", c.ID, "attempt", attempted+1, "error", startErr)
 
-	// If it is a 500 error it is likely we can retry and be successful
-	if strings.Contains(startErr.Error(), "API error (500)") {
+	if isDockerTransientError(startErr) {
 		if attempted < 5 {
 			attempted++
 			time.Sleep(1 * time.Second)
```
Comment (nickethier) [inline drivers/docker/driver.go]: should probably backoff here as well

File: drivers/docker/driver.go (hunk 1 above)
Comment (notnoop) [reply drivers/docker/driver.go]: Sounds good.  Makes sense for more aggressive refactor.

**Pull request merged**